### PR TITLE
MLIR MIOpen Dialect integration (phase 1) (#768)

### PR DESCRIPTION
**Pull request by sjw36**
* Added Findmlir.cmake (using environment variables to import)

* Added mlir_conv pass to GPU target

  * Apply to any gpu::convolution if supported by MLIR

  * Call MLIR C-API to generate iGEMM kernel with configuration from gpu::convolution

  * Capture binary in dictionary for matching convolutions

  * Build a code_object_op with the binary and execution dimensions

  * Substitute for the gpu::convolution

### CODE REVIEW

**Comment by sjw36** [comment]
@pfultz2 this enables MLIR substitution but I am getting a crash in allocating params:

> Allocating params ...
> Memory access fault by GPU node-4 (Agent handle: 0x3338700) on address 0x1000. Reason: Page not present or supervisor privilege.
> Aborted (core dumped)

It will also require a local build of https://github.com/ROCmSoftwarePlatform/llvm-project-mlir once this PR is merged (https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/pull/64).

File: cmake/Findmlir.cmake
Comment (pfultz2) [inline cmake/Findmlir.cmake:7]: We shouldn't rely on env variables to find dependencies. This should just use cmake's find_* command to find the different pieces needed. Ideally, it would be even nicer if mlir could provide a mlir-config.cmake so we dont need to this find module, but this is fine for a prototyping phase.

File: src/targets/gpu/mlir_conv.cpp
Comment (pfultz2) [inline src/targets/gpu/mlir_conv.cpp:211]: This follow the lowering passes, but I think it makes it more complicated in this case. You could use the matchers:

```cpp
struct find_mlir_convolution
{
    context * ctx = nullptr;
    auto matcher() const
    {
        return match::name("convolution");
    }

    void apply(module& m, match::matcher_result r) const
    {
        auto ins = r.result;
        // Replace ins with gpu::convolution
        ...
    }
};

void target_mlir_conv::apply(module& m) const 
{ 
    find_matches(m, find_mlir_convolution{&ctx});
}
```

Which could extend this to match more operators together(like conv+bias+activation).

However, if you just want to look at convolution, an even simpler way is to just use a simple for loop:

```cpp
void target_mlir_conv::apply(module& m) const 
{ 
    for(auto ins:iterator_for(m))
    {
        if (ins->name() != "convolution")
            continue;
        // Replace ins with gpu::convolution
        ...
    }
}
```

File: src/targets/gpu/CMakeLists.txt
Comment (pfultz2) [inline src/targets/gpu/CMakeLists.txt:257]: We should have an explicit cmake flag to enable MLIR: `set(MIGRAPHX_ENABLE_MLIR Off CACHE BOOL "")` and then only add those flags if it is set.

File: src/targets/gpu/CMakeLists.txt
Comment (pfultz2) [inline src/targets/gpu/CMakeLists.txt:258]: This should search for the library with `find_library`, similar to how MIOpen finds MLIR:

https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/CMakeLists.txt#L318

File: src/targets/gpu/mlir_conv.cpp
Comment (pfultz2) [inline src/targets/gpu/mlir_conv.cpp:65]: I think this should just return by value. We know if it didn't create the binary if `binary.empty()` is true.

File: src/targets/gpu/mlir_conv.cpp
Comment (pfultz2) [inline src/targets/gpu/mlir_conv.cpp:140]: The handle should be created with a managed pointer(which is just a `std::unique_ptr` with a custom deleter) so it will cleanup even if an exception occurs:

```cpp
using mlir_handle = MIGRAPHX_MANAGE_PTR(MiirHandle, miirDestroyHandle);

auto handle = mlir_handle{mlir_options.c_str()};
```

File: src/targets/gpu/mlir_conv.cpp
Comment (pfultz2) [inline src/targets/gpu/mlir_conv.cpp:144]: This is never deleted, use [`make_shared_array`](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/include/migraphx/make_shared_array.hpp#L11) here.

File: src/targets/gpu/mlir_conv.cpp
Comment (pfultz2) [inline src/targets/gpu/mlir_conv.cpp:210]: This can be written with std::transform instead.

File: src/targets/gpu/mlir_conv.cpp
Comment (sjw36) [reply src/targets/gpu/mlir_conv.cpp:65]: That would induce a copy, would be very expensive with the binary vector copy.

Comment (pfultz2) [reply src/targets/gpu/mlir_conv.cpp:65]: NRVO should elide the copy, but even if that is not used, the compiler should implicitly move the value in C++14(according to [1579](http://open-std.org/JTC1/SC22/WG21/docs/cwg_defects.html#1579)), which will avoid a copy, and just do some pointer swaps.

Comment (sjw36) [reply src/targets/gpu/mlir_conv.cpp:65]: But the execution_spec is potentially shared from the unordered_map (which owns the object). That way multiple conv2d's with the same config will not call the compiler every time.

Comment (pfultz2) [reply src/targets/gpu/mlir_conv.cpp:65]: I see, so if the map owns it, it should be `std::unordered_map<std::string, execution_spec>` and then this function can return a pointer to the element in the map(it doesn't seem like shared ownership is necessary), but you will need to add elements to the map with `binary_map.emplace()` so it can move(rather than copy) the object into the map.

Comment (sjw36) [reply src/targets/gpu/mlir_conv.cpp:65]: True. will do.

Comment (sjw36) [reply src/targets/gpu/mlir_conv.cpp:65]: Will use shared_ptr instead since we still want nullptr's in the map.

File: src/targets/gpu/mlir_conv.cpp
Comment (sjw36) [reply src/targets/gpu/mlir_conv.cpp:140]: How does that call miirCreateHandle?

Comment (pfultz2) [reply src/targets/gpu/mlir_conv.cpp]: Oops, that should be `auto handle = mlir_handle{miirCreateHandle(mlir_options.c_str())};`

File: src/targets/gpu/mlir_conv.cpp
Comment (pfultz2) [reply src/targets/gpu/mlir_conv.cpp]: Why do we want nullpointers in the map? Whats the difference between the key not being in the map and the key being in the map but the value is null?

File: Dockerfile
Comment (pfultz2) [inline Dockerfile:16]: Why are we installing clang?

File: src/targets/gpu/mlir_conv.cpp
Comment (pfultz2) [inline src/targets/gpu/mlir_conv.cpp:229]: It would be better to use implicit `[&]` capture for the lambda.

File: src/targets/gpu/mlir_conv.cpp
Comment (pfultz2) [inline src/targets/gpu/mlir_conv.cpp:231]: This can be done with `std::transform` using `std::back_inserter`:

```cpp
std::transform(inst_t.lens().begin(), inst_t.lens().end(), std::back_inserter(refs), [&](const auto& lval) {
    return get_literal(lval);
});
```

File: src/targets/gpu/mlir_conv.cpp
Comment (pfultz2) [inline src/targets/gpu/mlir_conv.cpp:271]: This can be written as `for(auto it:iterator_for(*prog))`

File: src/targets/gpu/mlir_conv.cpp
Comment (pfultz2) [inline src/targets/gpu/mlir_conv.cpp:38]: The variable should be named `mod` and not `prog` as its a module and not a program.

File: src/targets/gpu/mlir_conv.cpp
Comment (pfultz2) [inline src/targets/gpu/mlir_conv.cpp:281]: It would be better for the parameter to be `m`(for module) instead of `p`(which would stand for `program`).

File: src/targets/gpu/mlir_conv.cpp
Comment (pfultz2) [reply src/targets/gpu/mlir_conv.cpp]: Wait, I see, null means it tried but failed to compile a kernel, whereas a missing key means it hasn't even tried.

File: src/include/migraphx/value.hpp
Comment (pfultz2) [inline src/include/migraphx/value.hpp:143]: Is there a reason you can't cast it to `char*` like what we do for [msgpack](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/msgpack.cpp#L88)?

This also breaks the unit tests.

File: Dockerfile
Comment (sjw36) [reply Dockerfile:16]: No reason. Removing..

File: src/targets/gpu/mlir_conv.cpp
Comment (sjw36) [reply src/targets/gpu/mlir_conv.cpp:231]: Cool, I will try that.

File: src/include/migraphx/value.hpp
Comment (sjw36) [reply src/include/migraphx/value.hpp:143]: No problem, I tried this first but didn't work so I added the cast and forgot to remove this.

File: src/targets/gpu/mlir_conv.cpp
Comment (sjw36) [reply src/targets/gpu/mlir_conv.cpp:229]: We prefer explicit capture in MLIR but I can match your preference here.

File: src/targets/gpu/CMakeLists.txt
Comment (pfultz2) [inline src/targets/gpu/CMakeLists.txt:264]: This should be set to `Off` by default.

File: src/include/migraphx/compile_options.hpp
Comment (pfultz2) [inline src/include/migraphx/compile_options.hpp:14]: I think for now we shouldn't need a flag. By default, migraphx doesn't use MLIR so this flag seems unnecessary. We can use the `MIGRAPHX_ENABLE_MLIR` build flag to decide if its enabled or not.

The problem with a flag like this is that none of the verification tests enable it. With a build time flag we can easily add a Jenkins stage with `-DMIGRAPHX_ENABLE_MLIR=On` and it will run all the verification tests with mlir enable. Once we get further along to use MLIR by default, then we can think about how we might want to enable or disable it.

**Comment by pfultz2** [comment]
We can add a stage to jenkins after `clang_release` to run and build mlir:

```groovy
rocmtest clang_debug: rocmnode('vega') { cmake_build ->
    stage('Hip Clang Debug') {
        def sanitizers = "undefined"
        def debug_flags = "-g -O2 -fsanitize=${sanitizers} -fno-sanitize-recover=${sanitizers}"
        cmake_build("/opt/rocm/llvm/bin/clang++", "-DCMAKE_BUILD_TYPE=debug -DMIGRAPHX_ENABLE_PYTHON=Off -DCMAKE_CXX_FLAGS_DEBUG='${debug_flags}'")
    }
}, clang_release: rocmnode('vega') { cmake_build ->
    stage('Hip Clang Release') {
        cmake_build("/opt/rocm/llvm/bin/clang++", "-DCMAKE_BUILD_TYPE=release")
        stash includes: 'build/*.deb', name: 'migraphx-package'
    }
}, mlir_debug: rocmnode('vega') { cmake_build ->
    stage('MLIR Debug') {
        def sanitizers = "undefined"
        def debug_flags = "-g -O2 -fsanitize=${sanitizers} -fno-sanitize-recover=${sanitizers}"
        cmake_build("/opt/rocm/llvm/bin/clang++", "-DCMAKE_BUILD_TYPE=debug -DMIGRAPHX_ENABLE_PYTHON=Off -DMIGRAPHX_ENABLE_MLIR=On -DCMAKE_CXX_FLAGS_DEBUG='${debug_flags}'")
    }
}
```

File: src/include/migraphx/compile_options.hpp
Comment (sjw36) [reply src/include/migraphx/compile_options.hpp]: gone.

**Comment by sjw36** [comment]
I believe all issues are addressed. Anything else @pfultz2 ?

**Comment by pfultz2** [comment]
There is still an issue on jenkins:

```
/var/jenkins/workspace/MLLibs_AMDMIGraphX_PR-769/src/targets/gpu/mlir_conv.cpp:71:70: error: unused parameter 'op_r' [-Werror,-Wunused-parameter]
```

Its really only unused when MLIR is not enabled, so you should probably add a `(void)op_r` line on the `#else` part of the ifdef.

Also, can you add a stage to jenkins with MLIR enabled? I mentioned how [here](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/pull/769#issuecomment-825887677).

**Comment by pfultz2** [comment]
Also, it looks formatting is needed again:

```
sh -c clang-format-5.0 -style=file ./src/targets/gpu/mlir_conv.cpp | diff - ./src/targets/gpu/mlir_conv.cpp 
285c285,288
< void mlir_conv::apply(module& m) const { mlir_apply{&m, this}.apply(); }
---
> void mlir_conv::apply(module& m) const
> {
>     mlir_apply{&m, this}.apply();
> }
```

**Comment by sjw36** [comment]
what is up with the false error on deepcode?

**Review by pfultz2 — APPROVED** [review]

**Comment by sjw36** [comment]
still blocked..

**Comment by pfultz2** [comment]
There's a new compiler warning from merging in the tuple type PR:

```
[2021-04-28T01:15:33.019Z] /var/jenkins/workspace/MLLibs_AMDMIGraphX_PR-769/src/targets/gpu/mlir_conv.cpp:82:20: error: enumeration value 'tuple_type' not handled in switch [-Werror,-Wswitch]

[2021-04-28T01:15:33.019Z]             switch(s.type())

[2021-04-28T01:15:33.019Z]                    ^
```

**Comment by sjw36** [comment]
@mvermeulen looks like it passed all required. I don't have merge permissions, please merge. thx

**Pull request merged**
